# 전자제조데이터분석 과제 03
## US - Baby Names

아래 각 스텝의 결과물을 제시할 수 있도록 코드를 완성하세요.
본 과제 노트북 파일을 압축하여 eclass 과제 03 메뉴로 제출하세요


**Due: 2022.10.16 23:59:59**

학번: 22510107

이름: 이다인



### Step 1. pandas 모듈을 import 후 버전 정보를 출력하세요


In [1]:
import pandas as pd
pd.__version__

'1.3.5'

In [ ]:
# 출력 예시는 아래와 같음

### Step 2. 제공된 csv 파일들을 Colab 파일 시스템으로 업로드하세요 (어떠한 방법을 사용해도 상관 없음)

### Step 3. pandas 의 read_csv 함수를 이용하여 US_baby.csv 파일을 DataFrame 형태로 읽어들이고 baby_names 라는 변수에 할당하세요

In [19]:
baby_names = pd.read_csv("US_Baby.csv")

In [3]:
baby_names.info() # 아래와 같은 출력이 나와야 함
# Name: 아이의 이름
# Year: 출생연도
# Gender: 성별
# State: 출생지역 (주)
# Count: 해당 이름을 가진 아이의 수

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1016395 entries, 0 to 1016394
Data columns (total 7 columns):
 #   Column      Non-Null Count    Dtype 
---  ------      --------------    ----- 
 0   Unnamed: 0  1016395 non-null  int64 
 1   Id          1016395 non-null  int64 
 2   Name        1016395 non-null  object
 3   Year        1016395 non-null  int64 
 4   Gender      1016395 non-null  object
 5   State       1016395 non-null  object
 6   Count       1016395 non-null  int64 
dtypes: int64(4), object(3)
memory usage: 54.3+ MB


### Step 4. baby_names의 첫 10번째 행 정보를 출력하세요

In [20]:
baby_names.head(10)

Unnamed: 0     Id      Name  Year Gender State  Count
0       11349  11350      Emma  2004      F    AK     62
1       11350  11351   Madison  2004      F    AK     48
2       11351  11352    Hannah  2004      F    AK     46
3       11352  11353     Grace  2004      F    AK     44
4       11353  11354     Emily  2004      F    AK     41
5       11354  11355   Abigail  2004      F    AK     37
6       11355  11356    Olivia  2004      F    AK     33
7       11356  11357  Isabella  2004      F    AK     30
8       11357  11358    Alyssa  2004      F    AK     29
9       11358  11359    Sophia  2004      F    AK     28

### Step 5. baby_names에서 'Unnamed 0'과 'ID' 컬럼을 삭제하고 그 결과를 확인하세요.

In [21]:
baby_names.drop(labels = ['Unnamed: 0', 'Id'], axis=1).head(10)

Name  Year Gender State  Count
0      Emma  2004      F    AK     62
1   Madison  2004      F    AK     48
2    Hannah  2004      F    AK     46
3     Grace  2004      F    AK     44
4     Emily  2004      F    AK     41
5   Abigail  2004      F    AK     37
6    Olivia  2004      F    AK     33
7  Isabella  2004      F    AK     30
8    Alyssa  2004      F    AK     29
9    Sophia  2004      F    AK     28

### Step 6. 남성과 여성 중 어떤 성별의 아이가 많은지 알아보세요. 또한, 그 근거를 제시하세요.

In [26]:
g1 = baby_names.groupby('Gender').sum()
g1['Count']

Gender
F    16380293
M    19041199
Name: Count, dtype: int64

In [8]:
# print("female : ", len(baby_names[baby_names['Gender']=='F']))
# print("male : ", len(baby_names[baby_names['Gender']=='M']))

-> 여자 아이가 16380293명, 남자 아이가 19041199명으로 남자 아이가 더 많다.

---



### Step 7. baby_names 데이터셋에는 몇 개의 서로 다른 이름이 존재하는지 제시하세요.

In [27]:
name = set(baby_names['Name'].values.tolist())
print(len(name))

17632


### Step 8. 가장 많이 등장한 이름은 무엇인가요? 아이의 이름과 등장 횟수를 제시하세요.

In [28]:
g2 = baby_names.groupby("Name").sum('Count')
g2.sort_values('Count', ascending=False)

Unnamed: 0          Id     Year   Count
Name                                               
Jacob       1665680788  1665681356  1141099  242874
Emma        1629481684  1629482250  1137085  214852
Michael     1687520717  1687521295  1161152  214405
Ethan       1660807908  1660808475  1139091  209277
Isabella    1630131220  1630131786  1137090  204798
...                ...         ...      ...     ...
Eniola         4930125     4930126     2012       5
Atlantis       4904426     4904427     2005       5
Marci          4901272     4901273     2004       5
Simarpreet      553385      553386     2011       5
Nita            536781      536782     2007       5

[17632 rows x 4 columns]

-> Name으로 그룹을 만들어서 같은 이름의 아이가 몇 명인지 count를 반영하여 sum을 계산한다.

-> 가장 많이 등장한 이름은 Jacob로 총 242874번 등장했다.

---



### Step 9. 주어진 state.csv 파일을 업로드하고 pandas 의 read_csv 함수를 이용하여 DataFrame 형태로 읽어들이고 states 라는 변수에 할당하세요

In [29]:
states = pd.read_csv("state.csv")

In [30]:
states.info()
# 아래와 같은 정보가 출력되어야 함
# State: 주의 fullname
# Abbrev: 주의 약어
# Code: 주의 code

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   State   51 non-null     object
 1   Abbrev  51 non-null     object
 2   Code    51 non-null     object
dtypes: object(3)
memory usage: 1.3+ KB


In [31]:
states.head()

State  Abbrev Code
0     Alabama    Ala.   AL
1      Alaska  Alaska   AK
2     Arizona   Ariz.   AZ
3    Arkansas    Ark.   AR
4  California  Calif.   CA

# Step 10. states dataframe에는 각 주의 Code, abbreviation, fullname이 포함되어 있습니다. states와 baby_names dataframe을 적절히 병합하여 baby_names 내 각 주의 코드명을 fullname으로 변경하세요.

In [32]:
baby_names = pd.read_csv('US_Baby.csv')

In [33]:
baby = baby_names.drop(labels = ['Unnamed: 0', 'Id'], axis=1)

In [34]:
baby_names = pd.merge(baby, states, left_on = 'State', right_on='Code').drop(['Code', 'Abbrev'], axis=1)

In [35]:
baby_names = baby_names.drop(labels = 'State_x', axis=1)
baby_names.columns = ["Name", "Year", "Gender", "Count", "State"]

In [36]:
baby_names

Name  Year Gender  Count    State
0           Emma  2004      F     62   Alaska
1        Madison  2004      F     48   Alaska
2         Hannah  2004      F     46   Alaska
3          Grace  2004      F     44   Alaska
4          Emily  2004      F     41   Alaska
...          ...   ...    ...    ...      ...
1016390     Seth  2014      M      5  Wyoming
1016391  Spencer  2014      M      5  Wyoming
1016392     Tyce  2014      M      5  Wyoming
1016393   Victor  2014      M      5  Wyoming
1016394   Waylon  2014      M      5  Wyoming

[1016395 rows x 5 columns]

In [ ]:
baby_names
#출력 예시는 아래와 같음

Name  Year Gender  Count    State
0           Emma  2004      F     62   Alaska
1        Madison  2004      F     48   Alaska
2         Hannah  2004      F     46   Alaska
3          Grace  2004      F     44   Alaska
4          Emily  2004      F     41   Alaska
...          ...   ...    ...    ...      ...
1016390     Seth  2014      M      5  Wyoming
1016391  Spencer  2014      M      5  Wyoming
1016392     Tyce  2014      M      5  Wyoming
1016393   Victor  2014      M      5  Wyoming
1016394   Waylon  2014      M      5  Wyoming

[1016395 rows x 5 columns]

# Step 11. 아이가 가장 많이 태어난 년도는 언제인가요? 정답과 근거를 제시하세요.

In [40]:
count_year = baby_names.groupby('Year').sum()
count_year.sort_values('Count', ascending=False)

Count
Year         
2007  3412417
2006  3387737
2008  3340318
2005  3305315
2004  3293984
2009  3232116
2010  3116548
2014  3113611
2011  3079145
2012  3073858
2013  3066443

-> year를 기준으로 그룹을 묶어서 count에 대해 sum을 수행해 각 년도에 얼마나 아이가 태어났는지를 확인할 수 있다.

-> 2007년에 가장 많은 아이가 태어났다.




#Step 12. 미국에는 몇 명의 천둥의신이 존재하는지 제시하세요. 즉, 이름이 토르(Thor)인 아이들의 수를 세어보세요.

In [41]:
name_Thor = baby_names[baby_names['Name']=="Thor"]
name_Thor.head()

Name  Year Gender  Count       State
61990   Thor  2014      M      8     Arizona
109440  Thor  2004      M      7  California
111382  Thor  2005      M     14  California
114805  Thor  2006      M      8  California
118265  Thor  2007      M      6  California

In [42]:
name_Thor['Count'].sum()

419

-> baby_names에서 Name 변수가 Thor인 부분만 가져와서

-> Count 만큼을 다 더해야 이름이 Thor인 아이들의 수를 구할 수 있다.

-> 따라서 Count에 대해 sum을 수행하면 419명 을 얻을 수 있다.

# Step 13. 각 주의 년도별 출생아수를 아래와 같이 출력하세요.

In [ ]:
baby_names.groupby(['State', 'Year'])['Count'].sum().unstack()

In [43]:
g3 = baby_names.groupby(['State', 'Year'],as_index=True).sum()

In [44]:
g3

Count
State   Year       
Alabama 2004  42874
        2005  43405
        2006  45589
        2007  46346
        2008  45992
...             ...
Wyoming 2010   3162
        2011   2989
        2012   3051
        2013   3193
        2014   3151

[561 rows x 1 columns]

In [45]:
stateyear = g3.unstack(level=-1)
stateyear

Count                                                  \
Year                    2004    2005    2006    2007    2008    2009    2010   
State                                                                          
Alabama                42874   43405   45589   46346   45992   44063   42172   
Alaska                  5588    5614    5719    5779    5931    5796    5832   
Arizona                75582   77714   83222   82836   79484   72963   68327   
Arkansas               27044   27323   28042   28235   27348   26708   25618   
California            480593  484148  494605  497195  482863  459747  443930   
Colorado               54280   54647   55523   54830   54075   52473   50242   
Connecticut            31724   31003   30567   30395   29148   27881   26845   
Delaware                7180    7154    7076    7089    6828    6503    6277   
District of Columbia    8137    7513    7964    7863    7430    7241    6869   
Florida               177483  184536  192219  193135  186264  177253  171811   
Georgia               108781  110992  116136  117142  113797  109539  103552   
Hawaii                  8674    8087    8687    8317    8536    7966    8024   
Idaho                  15109   15431   16097   16455   16539   15507   14745   
Illinois              146721  144181  144990  144770  140755  135519  130380   
Indiana                71144   70639   71705   72153   70583   68776   66176   
Iowa                   29706   30047   30933   30438   29905   28905   28012   
Kansas                 30079   30136   30581   31171   30650   30391   29598   
Kentucky               41341   43064   44179   44527   43689   42560   40654   
Louisiana              46250   43758   45065   47032   45323   45623   43434   
Maine                  10029    9951    9777    9767    9207    8963    8577   
Maryland               52991   53524   55223   55765   55133   53438   52485   
Massachusetts          63303   61327   61304   60544   59188   57231   55499   
Michigan              104151  101877  101123   98840   95523   91709   89363   
Minnesota              56121   55916   57558   57122   55492   53893   51876   
Mississippi            26643   26297   28505   29154   28159   26607   24746   
Missouri               62127   62827   64447   64414   63652   61154   59696   
Montana                 6807    6745    7329    6988    7248    6940    6767   
Nebraska               18918   18576   19013   18989   18600   18544   17682   
Nevada                 24611   26217   28096   28681   27204   25734   24495   
New Hampshire          10315   10011    9954    9758    9514    9019    8895   
New Jersey             89048   87489   87713   87994   84831   82854   79927   
New Mexico             19000   19207   19761   19964   19612   18556   17425   
New York              206545  203136  205806  208233  204356  202135  198882   
North Carolina         95366   97665  101522  104197  103690  100146   96347   
North Dakota            5841    5936    5988    6109    6178    5950    6216   
Ohio                  123813  122494  123414  122729  120459  116730  111833   
Oklahoma               37457   37237   38946   39463   38725   38466   37305   
Oregon                 35212   35305   36805   37163   36668   34896   33318   
Pennsylvania          118657  118409  119651  120587  118954  116447  113557   
Rhode Island            8968    8699    8371    8217    7730    7275    7011   
South Carolina         38062   38604   41559   41846   42124   40179   38613   
South Dakota            7269    7386    7406    7530    7282    7128    6974   
Tennessee              66267   67785   69077   70948   69658   66580   64165   
Texas                 338179  341364  352121  359048  356351  352470  337544   
Utah                   40196   40718   42061   43319   43349   41897   40108   
Vermont                 3854    3772    3562    3573    3338    3184    3089   
Virginia               79612   80122   81778   82699   80478   78749   77108   
Washington             

In [ ]:
stateyear
#출력 예시는 아래와 같음

Count                                                  \
Year                    2004    2005    2006    2007    2008    2009    2010   
State                                                                          
Alabama                42874   43405   45589   46346   45992   44063   42172   
Alaska                  5588    5614    5719    5779    5931    5796    5832   
Arizona                75582   77714   83222   82836   79484   72963   68327   
Arkansas               27044   27323   28042   28235   27348   26708   25618   
California            480593  484148  494605  497195  482863  459747  443930   
Colorado               54280   54647   55523   54830   54075   52473   50242   
Connecticut            31724   31003   30567   30395   29148   27881   26845   
Delaware                7180    7154    7076    7089    6828    6503    6277   
District of Columbia    8137    7513    7964    7863    7430    7241    6869   
Florida               177483  184536  192219  193135  186264  177253  171811   
Georgia               108781  110992  116136  117142  113797  109539  103552   
Hawaii                  8674    8087    8687    8317    8536    7966    8024   
Idaho                  15109   15431   16097   16455   16539   15507   14745   
Illinois              146721  144181  144990  144770  140755  135519  130380   
Indiana                71144   70639   71705   72153   70583   68776   66176   
Iowa                   29706   30047   30933   30438   29905   28905   28012   
Kansas                 30079   30136   30581   31171   30650   30391   29598   
Kentucky               41341   43064   44179   44527   43689   42560   40654   
Louisiana              46250   43758   45065   47032   45323   45623   43434   
Maine                  10029    9951    9777    9767    9207    8963    8577   
Maryland               52991   53524   55223   55765   55133   53438   52485   
Massachusetts          63303   61327   61304   60544   59188   57231   55499   
Michigan              104151  101877  101123   98840   95523   91709   89363   
Minnesota              56121   55916   57558   57122   55492   53893   51876   
Mississippi            26643   26297   28505   29154   28159   26607   24746   
Missouri               62127   62827   64447   64414   63652   61154   59696   
Montana                 6807    6745    7329    6988    7248    6940    6767   
Nebraska               18918   18576   19013   18989   18600   18544   17682   
Nevada                 24611   26217   28096   28681   27204   25734   24495   
New Hampshire          10315   10011    9954    9758    9514    9019    8895   
New Jersey             89048   87489   87713   87994   84831   82854   79927   
New Mexico             19000   19207   19761   19964   19612   18556   17425   
New York              206545  203136  205806  208233  204356  202135  198882   
North Carolina         95366   97665  101522  104197  103690  100146   96347   
North Dakota            5841    5936    5988    6109    6178    5950    6216   
Ohio                  123813  122494  123414  122729  120459  116730  111833   
Oklahoma               37457   37237   38946   39463   38725   38466   37305   
Oregon                 35212   35305   36805   37163   36668   34896   33318   
Pennsylvania          118657  118409  119651  120587  118954  116447  113557   
Rhode Island            8968    8699    8371    8217    7730    7275    7011   
South Carolina         38062   38604   41559   41846   42124   40179   38613   
South Dakota            7269    7386    7406    7530    7282    7128    6974   
Tennessee              66267   67785   69077   70948   69658   66580   64165   
Texas                 338179  341364  352121  359048  356351  352470  337544   
Utah                   40196   40718   42061   43319   43349   41897   40108   
Vermont                 3854    3772    3562    3573    3338    3184    3089   
Virginia               79612   80122   81778   82699   80478   78749   77108   
Washington             

# Step 14. baby_names에 'birth_diff'라는 column을 추가하여 2004년도 대비 2014년도의 출생아 비율을 계산하여 추가하세요. 

#그 후, 2004년도 대비 2014년도 출생아 비율이 1을 초과하는 주에 대한 정보만 출력하세요.
- 예를 들어, 아래 출력과 같이 Alaska의 경우, 2004년도에는 5588명, 2014년도에는 5955명이므로 birth_diff는 5955/5588인 1.065676으로 계산될 수 있습니다.


In [46]:
stateyear[("Count", 2014)]

State
Alabama                  42139
Alaska                    5955
Arizona                  67808
Arkansas                 25799
California              436552
Colorado                 50102
Connecticut              26609
Delaware                  6356
District of Columbia      7607
Florida                 178147
Georgia                 102270
Hawaii                    7877
Idaho                    14498
Illinois                124514
Indiana                  65913
Iowa                     28309
Kansas                   28541
Kentucky                 40034
Louisiana                45556
Maine                     8145
Maryland                 52173
Massachusetts            54880
Michigan                 89460
Minnesota                52365
Mississippi              24834
Missouri                 58808
Montana                   6904
Nebraska                 18232
Nevada                   23893
New Hampshire             8279
New Jersey               78544
New Mexico               15737
Ne

In [47]:
stateyear['birth_diff'] = stateyear[("Count", 2014)] / stateyear[("Count", 2004)]

In [48]:
stateyear

Count                                                  \
Year                    2004    2005    2006    2007    2008    2009    2010   
State                                                                          
Alabama                42874   43405   45589   46346   45992   44063   42172   
Alaska                  5588    5614    5719    5779    5931    5796    5832   
Arizona                75582   77714   83222   82836   79484   72963   68327   
Arkansas               27044   27323   28042   28235   27348   26708   25618   
California            480593  484148  494605  497195  482863  459747  443930   
Colorado               54280   54647   55523   54830   54075   52473   50242   
Connecticut            31724   31003   30567   30395   29148   27881   26845   
Delaware                7180    7154    7076    7089    6828    6503    6277   
District of Columbia    8137    7513    7964    7863    7430    7241    6869   
Florida               177483  184536  192219  193135  186264  177253  171811   
Georgia               108781  110992  116136  117142  113797  109539  103552   
Hawaii                  8674    8087    8687    8317    8536    7966    8024   
Idaho                  15109   15431   16097   16455   16539   15507   14745   
Illinois              146721  144181  144990  144770  140755  135519  130380   
Indiana                71144   70639   71705   72153   70583   68776   66176   
Iowa                   29706   30047   30933   30438   29905   28905   28012   
Kansas                 30079   30136   30581   31171   30650   30391   29598   
Kentucky               41341   43064   44179   44527   43689   42560   40654   
Louisiana              46250   43758   45065   47032   45323   45623   43434   
Maine                  10029    9951    9777    9767    9207    8963    8577   
Maryland               52991   53524   55223   55765   55133   53438   52485   
Massachusetts          63303   61327   61304   60544   59188   57231   55499   
Michigan              104151  101877  101123   98840   95523   91709   89363   
Minnesota              56121   55916   57558   57122   55492   53893   51876   
Mississippi            26643   26297   28505   29154   28159   26607   24746   
Missouri               62127   62827   64447   64414   63652   61154   59696   
Montana                 6807    6745    7329    6988    7248    6940    6767   
Nebraska               18918   18576   19013   18989   18600   18544   17682   
Nevada                 24611   26217   28096   28681   27204   25734   24495   
New Hampshire          10315   10011    9954    9758    9514    9019    8895   
New Jersey             89048   87489   87713   87994   84831   82854   79927   
New Mexico             19000   19207   19761   19964   19612   18556   17425   
New York              206545  203136  205806  208233  204356  202135  198882   
North Carolina         95366   97665  101522  104197  103690  100146   96347   
North Dakota            5841    5936    5988    6109    6178    5950    6216   
Ohio                  123813  122494  123414  122729  120459  116730  111833   
Oklahoma               37457   37237   38946   39463   38725   38466   37305   
Oregon                 35212   35305   36805   37163   36668   34896   33318   
Pennsylvania          118657  118409  119651  120587  118954  116447  113557   
Rhode Island            8968    8699    8371    8217    7730    7275    7011   
South Carolina         38062   38604   41559   41846   42124   40179   38613   
South Dakota            7269    7386    7406    7530    7282    7128    6974   
Tennessee              66267   67785   69077   70948   69658   66580   64165   
Texas                 338179  341364  352121  359048  356351  352470  337544   
Utah                   40196   40718   42061   43319   43349   41897   40108   
Vermont                 3854    3772    3562    3573    3338    3184    3089   
Virginia               79612   80122   81778   82699   80478   78749   77108   
Washington             

In [50]:
stateyear[stateyear['birth_diff'] > 1]

Count                                                  \
Year              2004    2005    2006    2007    2008    2009    2010   
State                                                                    
Alaska            5588    5614    5719    5779    5931    5796    5832   
Florida         177483  184536  192219  193135  186264  177253  171811   
Montana           6807    6745    7329    6988    7248    6940    6767   
North Carolina   95366   97665  101522  104197  103690  100146   96347   
North Dakota      5841    5936    5988    6109    6178    5950    6216   
South Carolina   38062   38604   41559   41846   42124   40179   38613   
South Dakota      7269    7386    7406    7530    7282    7128    6974   
Tennessee        66267   67785   69077   70948   69658   66580   64165   
Texas           338179  341364  352121  359048  356351  352470  337544   
Washington       63237   64022   66909   68368   69271   68223   65794   
Wyoming           3082    3378    3512    3610    3534    3357    3162   

                                               birth_diff  
Year              2011    2012    2013    2014             
State                                                      
Alaska            5912    5631    5892    5955   1.065676  
Florida         171020  170912  173599  178147   1.003741  
Montana           6609    6612    6816    6904   1.014250  
North Carolina   94868   94283   94381   96215   1.008903  
North Dakota      6236    6623    6992    7492   1.282657  
South Carolina   38398   38183   38533   39258   1.031422  
South Dakota      6994    7042    7243    7293   1.003302  
Tennessee        64067   65204   64740   66611   1.005191  
Texas           329998  334705  339517  350378   1.036073  
Washington       65973   66162   65201   67161   1.062052  
Wyoming           2989    3051    3193    3151   1.022388

In [ ]:
# 출력 예시는 아래와 같음 

Count                                                  \
Year              2004    2005    2006    2007    2008    2009    2010   
State                                                                    
Alaska            5588    5614    5719    5779    5931    5796    5832   
Florida         177483  184536  192219  193135  186264  177253  171811   
Montana           6807    6745    7329    6988    7248    6940    6767   
North Carolina   95366   97665  101522  104197  103690  100146   96347   
North Dakota      5841    5936    5988    6109    6178    5950    6216   
South Carolina   38062   38604   41559   41846   42124   40179   38613   
South Dakota      7269    7386    7406    7530    7282    7128    6974   
Tennessee        66267   67785   69077   70948   69658   66580   64165   
Texas           338179  341364  352121  359048  356351  352470  337544   
Washington       63237   64022   66909   68368   69271   68223   65794   
Wyoming           3082    3378    3512    3610    3534    3357    3162   

                                               birth_diff  
Year              2011    2012    2013    2014             
State                                                      
Alaska            5912    5631    5892    5955   1.065676  
Florida         171020  170912  173599  178147   1.003741  
Montana           6609    6612    6816    6904   1.014250  
North Carolina   94868   94283   94381   96215   1.008903  
North Dakota      6236    6623    6992    7492   1.282657  
South Carolina   38398   38183   38533   39258   1.031422  
South Dakota      6994    7042    7243    7293   1.003302  
Tennessee        64067   65204   64740   66611   1.005191  
Texas           329998  334705  339517  350378   1.036073  
Washington       65973   66162   65201   67161   1.062052  
Wyoming           2989    3051    3193    3151   1.022388